src/MSMToolsCluster.jl

```diff
@@ -0,0 +1,259 @@
+using LinearAlgebra
+using Arpack
+using Statistics
+using SparseArrays
+using LightGraphs
+using Optim
+
+# NOTE: Reimplementation in Julia for select functions from MSM Tools (referenced in DeepTime)
+# https://github.com/markovmodel/msmtools
+# https://github.com/deeptime-ml/deeptime/tree/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/tools
+
+# Reimplementation of PCCA+ spectral clustering method with optimized memberships
+# https://github.com/deeptime-ml/deeptime/blob/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/tools/analysis/dense/_pcca.py#L288
+function pcca_memberships(P::Matrix{Float64}, m::Int, π::Vector{Float64}=nothing)
+    n = size(P, 1)
+    if m > n
+        throw(ArgumentError("Number of metastable states m=$m exceeds number of states of transition matrix n=$n"))
+    end
+    println("P: $P")
+    if !is_transition_matrix(P)
+        throw(ArgumentError("Input matrix is not a transition matrix."))
+    end
+    if isnothing(π)
+        if ndims(π) > 1
+            throw(ArgumentError("Expected 1D stationary distribution array"))
+        elseif length(π) != n
+            throw(ArgumentError("Stationary distribution must be defined on entire space, piecewise if the transition matrix " *
+            "has multiple connected components. It covered $(length(π)) != $n states."))
+        end
+    end
+
+    chi = zeros(n, m)
+    components = connected_sets(P)
+    closed_components = []
+    transition_states = []
```

Comment:
If the final size of an array and type are known it's best to instantiate an array beforehand. The empty arrays `[]` are of type `Array{Nothing}` which usually hurts performance significantly

Reply:
Great idea.. I have to double check that we definitely know the size of this ahead of time. If it's okay to leave this one for now, but I'll check the logic again asap.

Reply:
Even if the final size is not known, creating the arrays as `Float64[]` for instance, will avoid the issue of the arrays having elementy type `Any` which is boxed and thus much less efficient

Reply:
Ok, got it! Fixed now in bb1600c